pkg/controllers/static/helpers.go

```diff
@@ -0,0 +1,158 @@
+			IsStaticNode:      true,
+		})
+	}
+	return nodeClaims
+}
+
+func GetStaticNodeClaimTemplate(np *v1.NodePool, instanceTypes []*cloudprovider.InstanceType) *scheduling.NodeClaimTemplate {
+	nct := scheduling.NewNodeClaimTemplate(np)
+	nct.InstanceTypeOptions, _, _ = scheduling.FilterInstanceTypesByRequirements(
+		instanceTypes,
+		nct.Requirements,
+		corev1.ResourceList{},
+		corev1.ResourceList{},
+		corev1.ResourceList{},
+		true,
+	)
+	return nct
+}
+
+func ComputeNodeClaimsToProvision(c *state.Cluster, np *v1.NodePool, nodes int64) int64 {
+	limit, ok := np.Spec.Limits[resources.Node]
+	nodeLimit := lo.Ternary(ok, limit.Value(), int64(math.MaxInt64))
+	return c.NodePoolState.ReserveNodeCount(np.Name, nodeLimit, lo.FromPtr(np.Spec.Replicas)-nodes)
+}
+
+func TotalNodesForNodePool(c *state.Cluster, np *v1.NodePool) int64 {
+	running, deleting := c.NodePoolState.GetNodeCount(np.Name)
+	return int64(running + deleting)
+}
+
+// Returns nodes suitable for deprovisioning, prioritizing:
+// 1. Empty nodes (nodes with no pods or only DaemonSet pods)
+// 2. If more nodes needed, nodes with lowest disruption cost
+func GetDeprovisioningCandidates(ctx context.Context, kubeClient client.Client, np *v1.NodePool, nodes []*state.StateNode, count int, clk clock.Clock) []*state.StateNode {
+	// First get empty nodes
+	emptyNodes := lo.Filter(nodes, func(node *state.StateNode, _ int) bool {
+		pods, err := node.Pods(ctx, kubeClient)
+		if err != nil {
+			log.FromContext(ctx).WithValues("node", node.Name()).Error(err, "unable to list pods, treating as non-empty")
+			return false
+		}
+		return len(pods) == 0 || lo.EveryBy(pods, pod.IsOwnedByDaemonSet)
+	})
+
+	candidates := lo.Slice(emptyNodes, 0, count)
+	remaining := count - len(candidates)
+
+	if remaining > 0 {
+		emptyNodesSet := lo.SliceToMap(emptyNodes, func(n *state.StateNode) (*state.StateNode, struct{}) {
+			return n, struct{}{}
+		})
+		nonEmptyNodesWithCost := lo.FilterMap(nodes, func(node *state.StateNode, _ int) (NodeDisruptionCost, bool) {
+			if _, ok := emptyNodesSet[node]; ok {
+				return NodeDisruptionCost{}, false
+			}
+
+			pods, err := node.Pods(ctx, kubeClient)
+			if err != nil {
+				log.FromContext(ctx).WithValues("node", node.Name()).Error(err, "unable to list pods, skipping node")
+				return NodeDisruptionCost{}, false
+			}
+
+			return NodeDisruptionCost{
+				node: node,
+				cost: disruptionutils.ReschedulingCost(ctx, pods) *
+					disruptionutils.LifetimeRemaining(clk, np, node.NodeClaim),
+			}, true
+		})
+
+		// Optimization : Dont need to sort all items, instead use a heap to get remaining
+		sort.Slice(nonEmptyNodesWithCost, func(i, j int) bool {
+			return nonEmptyNodesWithCost[i].cost < nonEmptyNodesWithCost[j].cost
+		})
+
+		// Take the remaining needed nodes with lowest cost
+		lowestCostNodes := lo.Map(
+			lo.Slice(nonEmptyNodesWithCost, 0, remaining),
+			func(nwc NodeDisruptionCost, _ int) *state.StateNode {
+				return nwc.node
+			},
+		)
```

Comment:
I'd lean away from chaining `lo`, this has a lot of unneeded allocations. It's not too big of a deal here since we're not in a hotpath, but a solution like this avoids the extra allocations while being more readable (in my opinion):
```go
for i, nwc := range nonEmptyNodesWithCost {
    if i == remaining {
        break
    }
    candidates = append(candidates, nwc.node)
}